src/block/block_ckpt.c

```diff
@@ -653,17 +664,218 @@ __ckpt_process(WT_SESSION_IMPL *session, WT_BLOCK *block, WT_CKPT *ckptbase)
     return (ret);
 }
 
+/*
+ * __ckpt_valid_blk_mods --
+ *     Make sure that this set of block mods reflects the current valid backup identifiers. If so,
+ *     there is nothing to do. If not, free up old information and set it up for the current
+ *     information.
+ */
+static int
+__ckpt_valid_blk_mods(WT_SESSION_IMPL *session, WT_BLOCK_MODS *blk_mod, u_int i)
+{
+    WT_BLKINCR *blk;
+    bool free, setup;
+
+    blk = &S2C(session)->incr_backups[i];
+
+    /*
+     * Check the state of our block list array compared to the global one. There are
+     * several possibilities:
+     * - There is no global information for this index, nothing to do but free our resources.
+     * - We don't have any backup information locally. Set up our entry.
+     * - Our entry's id string matches the current global information. We just want to add our
+     *   information to the existing list.
+     * - Our entry's id string does not match the current one. It is outdated. Free old resources
+     *   and then set up our entry.
+     */
+
+    /*
+     * Check if the global entry is valid at our index.
+     */
+    if (!F_ISSET(blk, WT_BLKINCR_VALID)) {
+        free = true;
+        setup = false;
+    } else {
+        if (F_ISSET(blk_mod, WT_BLOCK_MODS_VALID) &&
+          WT_STRING_MATCH(blk_mod->id_str, blk->id_str, strlen(blk->id_str))) {
+            /* We match, keep our entry and don't set up. */
+            setup = false;
+            free = false;
+        } else {
+            /* We don't match, free any old information. */
+            free = true;
+            setup = true;
+        }
+    }
+
+    /*
+     * Free any old information if we need to do so.
+     */
+    if (free && F_ISSET(blk_mod, WT_BLOCK_MODS_VALID)) {
+        __wt_free(session, blk_mod->bitstring);
+        __wt_free(session, blk_mod->id_str);
+        blk_mod->nbits = 0;
+        blk_mod->granularity = 0;
+        F_CLR(blk_mod, WT_BLOCK_MODS_VALID);
+    }
+
+    /*
+     * Set up the block list to point to the current information.
+     */
+    if (setup) {
+        WT_RET(__wt_strdup(session, blk->id_str, &blk_mod->id_str));
+        blk_mod->nbits = 0;
+        blk_mod->bitstring = NULL;
+        blk_mod->granularity = S2C(session)->incr_granularity;
+        F_SET(blk_mod, WT_BLOCK_MODS_VALID);
+    }
+    return (0);
+}
+
+/*
+ * __ckpt_add_blkmod_entry --
+ *     Add an offset/length entry to the bitstring based on granularity.
+ */
+static int
+__ckpt_add_blkmod_entry(
+  WT_SESSION_IMPL *session, WT_BLOCK_MODS *blk_mod, uint64_t offset, uint64_t len)
+{
+    uint64_t end, start;
+    uint32_t end_rdup;
+
+    WT_ASSERT(session, blk_mod->granularity != 0);
+    start = offset / blk_mod->granularity;
+    end = (offset + len) / blk_mod->granularity;
```

Comment:
The granularity cannot change  for a single incremental once you've set it, but we save it on a per-incremental-id so it could change on a subsequent incremental. I think the API level does not allow it at the moment, but what we store and compute can be per-incremental.